Add support for synchronous FUSE_INIT

bsbernd opened the pull request.

bsbernd commented:
@szmi could you take a look, I think it works for privileged mounts. Will look into fusermount changes for unprivileged tomorrow.

szmi commented:
There's one situation that this doesn't handle well: if the fs crashes while processing the INIT request, the mount syscall will hang, because it's holding a ref to the device file.

Fixing this requires:

 - libfuse: switching to the new mount API
 - kernel: in fuse_parse_param() storing a reference to the fuse_conn instead of the fd
 - spawning a new process for mount, not just a new thread
 - after FSCONFIG_SET_FD and before FSCONFIG_CMD_CREATE closing the devfd

It's a bit more complicated, but may be worth it, otherwise any failure or crash during the initialization will result in a hang (resolvable only with with echo 1 > ../abort), which is a regression compared to the async mount case.

bsbernd committed:
Add fuse_session_set_debug() to enable debug output without foreground

One might want to print debug out in background mode, which the command
line parameter does not easily allow. Or one might want to enable and
disable at run time.

Signed-off-by: Bernd Schubert <bschubert@ddn.com>
Reviewed-by: "Darrick J. Wong" <djwong@kernel.org>

bsbernd committed:
Add a new daemonize API

Existing example/ file systems do the fuse_daemonize() after
fuse_session_mount() - i.e. after the mount point is already
established. Though, these example/ daemons do not start
extra threads and do not need network initialization either.

fuse_daemonize() also does not allow to return notification
from the forked child to the parent.

Complex fuse file system daemons often want the order of
1) fork - parent watches, child does the work

Child:
    2) start extra threads and system initialization (like network
       connection and RDMA memory registration) from the fork child.
    3) Start the fuse session after everything else succeeded

Parent:
    Report child initialization success or failure

A new API is introduced to overcome the limitations of
fuse_daemonize()

fuse_daemonize_start() - fork, but foreground process does not
terminate yet and watches the background.

fuse_daemonize_success() / fuse_daemonize_fail() - background
daemon signals to the foreground process success or failure.

fuse_daemonize_active() - helper function for the high level
interface, which needs to handle both APIs. fuse_daemonize()
is called within fuse_main(), which now needs to know if the caller
actually already used the new API itself.

The object 'struct fuse_daemonize *' is allocated dynamically
and stored a global variable in fuse_daemonize.c, because
- high level fuse_main_real_versioned() needs to know
if already daemonized
- high level daemons do not have access to struct fuse_session
- FUSE_SYNC_INIT in later commits can only be done if the new
API (or a file system internal) daemonization is used.

Signed-off-by: Bernd Schubert <bernd@bsbernd.com>

bsbernd committed:
Sync fuse_kernel.h with linux-6.18

Signed-off-by: Bernd Schubert <bschubert@ddn.com>
Reviewed-by: "Darrick J. Wong" <djwong@kernel.org>

bsbernd committed:
mount.c: Split fuse_mount_sys to prepare privileged sync FUSE_INIT

Synchronous FUSE_INIT needs to spawn a worker thread to handle
FUSE_INIT before starting the mount. For that it needs to have
the device file descriptor - this is preparation work and
fuse_mount_sys() is split into fuse_kern_mount_prepare()
and fuse_kern_do_mount().

Signed-off-by: Bernd Schubert <bschubert@ddn.com>
Reviewed-by: "Darrick J. Wong" <djwong@kernel.org>

bsbernd committed:
fuse mount: Support synchronous FUSE_INIT (privileged daemon)

Add synchronous FUSE_INIT processing during mount() to
enable early daemonization with proper error reporting
to the parent process.

A new mount thread is needed that handles FUSE_INIT and
possible other requests at mount time (like getxattr for selinux).
The kernel sends FUSE_INIT during the mount() syscall. Without a thread
to process it, mount() blocks forever.

Mount thread lifetime:
Created before mount() syscall in fuse_start_sync_init_worker()
Processes requests until se->mount_finished is set (after mount() returns)
Joined after successful mount in fuse_wait_sync_init_completion()
Cancelled if mount fails (direct → fusermount3 fallback)
Key changes:

Add init_thread, init_error, mount_finished to struct fuse_session
Use FUSE_DEV_IOC_SYNC_INIT ioctl for kernel support
Fall back to async FUSE_INIT if unsupported
Auto-enabled when fuse_daemonize_active() or via
fuse_session_want_sync_init()
Allows parent to report mount/init failures instead of
exiting immediately after fork.

Note: For now synchronous FUSE_INIT is only supported for privileged
mounts.

Signed-off-by: Bernd Schubert <bschubert@ddn.com>

bsbernd committed:
Add FUSE_MOUNT_FALLBACK_NEEDED define for -2 mount errors

Magic numbers in the code are not good - we better use a define.

Signed-off-by: Bernd Schubert <bschubert@ddn.com>
Reviewed-by: "Darrick J. Wong" <djwong@kernel.org>

bsbernd committed:
Refactor mount code / move common functions to mount_util.c

Also create the new "mount_i.h", which is independent of the
the rest of libfuse.

This is preparation for the new mount API, which goes into a new file.

Signed-off-by: Bernd Schubert <bschubert@ddn.com>
Reviewed-by: "Darrick J. Wong" <djwong@kernel.org>

bsbernd committed:
Add support for the new linux mount API

So far only supported for fuse_session_mount(), which is called
from high and low level API, but not yet supported for
fuse_open_channel(), which used for privilege drop through
mount.fuse. Main goal for the new API is support for synchronous
FUSE_INIT and I don't think that is going to work with
fuse_open_channel(). At least not with io-uring support as long
as it is started from FUSE_INIT.

Signed-off-by: Bernd Schubert <bschubert@ddn.com>

bsbernd committed:
checkpatch.pl: Add _Atomic to $Attribute patttern

WARNING: Missing a blank line after declarations
+       int init_error;
+       _Atomic bool terminate_mount_worker;

This change should go to linux upstream.

Signed-off-by: Bernd Schubert <bschubert@ddn.com>
Acked-by: "Darrick J. Wong" <djwong@kernel.org>

bsbernd committed:
conftest.py: Add more valgrind filter patterns

Valgrind complains that it does not know the fsmount syscall,
pytest checks for warnings, find that and fails the test

--14936-- WARNING: unhandled amd64-linux syscall: 430
--14936-- You may be able to write your own handler.
--14936-- Read the file README_MISSING_SYSCALL_OR_IOCTL.
--14936-- Nevertheless we consider this a bug.  Please report
--14936-- it at http://valgrind.org/support/bug_reports.html.
fuse: fsopen(fuse) failed: Function not implemented
=========================== short test summary info

Obviously we want to filter our valgrind warnings.

Signed-off-by: Bernd Schubert <bschubert@ddn.com>

bsbernd committed:
Move more generic mount code to mount_util.{c,h}

This is to allow fusermount to use the code from mount_fsmount.c.
I.e. avoid code dup and add just re-use the new linux api mount
functions from that file for fusermount.

Signed-off-by: Bernd Schubert <bschubert@ddn.com>
Reviewed-by: "Darrick J. Wong" <djwong@kernel.org>

bsbernd committed:
Split the fusermount do_mount function

We will need new API and old API and need to pass the options to
two different functions - factor out the option parsing.

Signed-off-by: Bernd Schubert <bschubert@ddn.com>
Reviewed-by: "Darrick J. Wong" <djwong@kernel.org>

bsbernd committed:
Make fusermount work bidirectional for sync init

Signed-off-by: Bernd Schubert <bschubert@ddn.com>

bsbernd committed:
fusermount: Refactor extract_x_options

Just to make it better readable.

Also add a NULL check for *regular_opts and *x_prefixed_opts,
there will be multiple callers in an upcoming commit and
it becomes harder to track that no caller has an error.

Signed-off-by: Bernd Schubert <bschubert@ddn.com>
Reviewed-by: "Darrick J. Wong" <djwong@kernel.org>

bsbernd committed:
New mount API: Filter out "user="

This gets added in the fusermount process and kernel then fails
the mount.

Signed-off-by: Bernd Schubert <bschubert@ddn.com>
Reviewed-by: "Darrick J. Wong" <djwong@kernel.org>

bsbernd committed:
ci-build: Add environment logging

It is very hard to see what exactly fails, as everything
is run from the same script.

Signed-off-by: Bernd Schubert <bschubert@ddn.com>
Acked-by: "Darrick J. Wong" <djwong@kernel.org>

bsbernd committed:
Add support for sync-init of unprivileged daemons

This makes use of the bidirectional fusermount. Added is
doc/README.mount, which explains the new bidirectional
communication with fusermount.

Signed-off-by: Bernd Schubert <bschubert@ddn.com>

bsbernd committed:
lib/mount.c: Remove some BSD ifdefs

BSD has its own mount_bsd.c, mount.c should be linux only.

Also remove MS_DIRSYNC defintion, which was introduced a long
time ago - all use cases of recent libfuse should have picked
it up.

Some ifdefs stay, as they will be moved in the next patch.

Suggested-by: "Darrick J. Wong" <djwong@kernel.org>
Signed-off-by: Bernd Schubert <bschubert@ddn.com>
Reviewed-by: "Darrick J. Wong" <djwong@kernel.org>

bsbernd committed:
fusermout: Remove the large read check

<quote Darrick>
Nobody's running 2.4 kernels
anymore, right?
</quote>

Signed-off-by: Bernd Schubert <bernd@bsbernd.com>
Reviewed-by: "Darrick J. Wong" <djwong@kernel.org>

bsbernd committed:
Move 'struct mount_flags' to util.h

We actually need these in fusermount.c and for the
new mount API, which goes into its own file.

Also extend the struct with the extra safe field
used by fusermount.c

Signed-off-by: Bernd Schubert <bernd@bsbernd.com>
Reviewed-by: "Darrick J. Wong" <djwong@kernel.org>

bsbernd committed:
Add mount and daemonization README documents

These are useful to

Signed-off-by: Bernd Schubert <bernd@bsbernd.com>

bsbernd committed:
Add a background debug option to passthrough hp

Background debugging is useful when logs should go
to syslog for a daemon running in background.

Signed-off-by: Bernd Schubert <bernd@bsbernd.com>

bsbernd committed:
Verify in fuse_mnt_get_devname() that the kern device is valid

codeql gave a warning that a user might pass a random device,
so we better check the device in /dev and the device is a character
device.

Signed-off-by: Bernd Schubert <bernd@bsbernd.com>

kchen-ddn commented on lib/mount.c:
```diff
@@ -524,50 +581,87 @@ static int fuse_mount_sys(const char *mnt, struct mount_opts *mo,
+	int res;
+	const char *devname = fuse_mnt_get_devname();
+	res = -ENOMEM;
+	source = fuse_mnt_build_source(mo->fsname, mo->subtype, devname);
```
The 'source' and 'type' may be changed below after mount failed with ENODEV, we'd better free  them and then call asprintf again for their modification.

bsbernd replied:
Good catch @kchen-ddn 
In the "Refactor mount code / move common functions to mount_util.c" patch it is still save - still uses malloc. Needs to be fixed in the commit that switches to asprintf.

bsbernd replied:
Actually it was safe, because subtype was just removed - the allocated array by asprintf() was guaranteed to be larger. But I still added the free and new asprintf() to ensure it cannot be broken in the future.

bsbernd committed:
Move fuse_daemonize() from helper.c to fuse_daemonize.c

The old API should be still in the generic daemonization file to
make it easier to find.

Also switch to err() to make checkpatch.pl happy.

Signed-off-by: Bernd Schubert <bernd@bsbernd.com>

bsbernd committed:
Add to mount_util / struct mount_flags if fsmount or fsconfig param

This is a preparation for the new mount API, add in if flags
are fsconfig or fsmount parameters.

Signed-off-by: Bernd Schubert <bernd@bsbernd.com>

bsbernd commented on example/passthrough_hp.cc:
```diff
@@ -243,6 +245,9 @@ static void sfs_init(void *userdata, fuse_conn_info *conn)
 	conn->max_write = 4 * 1024 * 1024;
+
+	/* required here for async init */
+	fuse_daemonize_success();
```
@kchen-ddn Removed at all in ->init() and done automatically at the end of `_do_init`

kchen-ddn commented on lib/mount_fsmount.c:
```diff
@@ -0,0 +1,477 @@
+ * Only handles flags that are filesystem parameters (ro, sync, dirsync).
+ * Mount attributes (nosuid, nodev, etc.) are handled separately via fsmount().
+ */
+static int set_ms_flags(int fsfd, unsigned long ms_flags)
```
This function never returns success.
The ms_flags can have other flags set, e.g. MS_NOSUID and MS_NODEV, etc.
In that case the for loop will iterate to the last fme of strflags, i.e. {0,0}.
Then 'fuse: set (null) option: Invalid argument' will be printed and then error returned.

Even if the ms_flags only contains MS_SYNCHRONOUS, MS_DIRSYNC and MS_LAZYTIME, after the for loop, it will return EOPNOTSUPP.

Should we only handle MS_SYNCHRONOUS, MS_DIRSYNC and MS_LAZYTIME in this function and leave the other flags to ms_flags_to_mount_attrs()?

bsbernd replied:
@kchen-ddn Thank you, that would have been a real issue. Given that tests did not find it, I need to add a test for it. Should be handled now, but I don't like the private maps in `mount_fsmount.c` yet. I think I'm going to add it to `const struct mount_flags mount_flags[]` in mount_utils.c

kchen-ddn replied:
Yes, the test could pass because it failed back to old API. I only noticed it because I was trying to do SELinux mount, which eventually hung at getxattr.

kchen-ddn commented on lib/mount.c:
I mean, here we do strcpy and sprintf to 'source' or 'type'. Should we free them and asprintf() again?

bsbernd replied:
I did that now, please check again, but it was actually safe, because subtype removal needs less space.

kchen-ddn replied:
Yes, thanks!

bsbernd committed:
Move fuse_mnt_build_{source,type} to mount_util.c

Also make it more generic and avoid taking struct mount_opts
as parameter.

mount_fsmount.c and fusermount.c use these functions now.

Also use asprintf as suggested by Darrick.
For safety also free source and type in fuse_kern_do_mount()
when subtype is not supported - right now it would be safe
because without subtype the allocated memory is just a bit too
large, but would easy to break in future commits.

Signed-off-by: Bernd Schubert <bernd@bsbernd.com>

kchen-ddn commented on lib/mount_fsmount.c:
```diff
@@ -0,0 +1,488 @@
+				fme->string, strerror(save_errno));
+			log_fsconfig_kmsg(fsfd);
+
+			return -save_errno;
```
This line we return an errno, but the return value should be the not handled flags.

bsbernd replied:
Thank you, updated. There was another issue, the loop did not check for map exhaustion - out of bound access.

bsbernd committed:
Unify mount flag structures and remove redundant is_mount_attr field

Refactor mount flag handling by merging struct ms_to_str_map and
struct ms_to_mount_attr_map (from mount_fsmount.c) into the existing
struct mount_flags (from mount_util.h/c).

This eliminates code duplication and provides a single source of truth
for all mount flag metadata, including:
- Option strings (e.g., "nosuid", "sync")
- MS_* flag values
- MOUNT_ATTR_* attribute values
- Classification (fsconfig vs fsmount)

Key changes:
- Added mount_attr field to struct mount_flags
- Removed redundant is_mount_attr field (replaced by mount_attr != 0)
- Updated mount_flags array with MOUNT_ATTR_* values
- Refactored ms_flags_to_mount_attrs() to use unified array
- Refactored set_ms_flags() to use unified array
- Updated is_mount_attr_opt() to check mount_attr field

Benefits:
- Easier maintenance when adding new mount flags
- Single consistent data structure for all mount APIs
- Clearer relationship between MS_* flags and MOUNT_ATTR_* attributes
- One less field in the structure (removed is_mount_attr)

MS_LAZYTIME that was defined before was not correct (at least
for now) and is removed again.

Signed-off-by: Bernd Schubert <bernd@bsbernd.com>
